src/core/SamplePreviewPlayHandle.cpp

```diff
@@ -0,0 +1,142 @@
+/*
+ * SamplePreviewPlayHandle.cpp
+ *
+ * Copyright (c) 2025 Sotonye Atemie <sakertooth@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SamplePreviewPlayHandle.h"
+
+#include "AudioBusHandle.h"
+#include "AudioEngine.h"
+#include "Engine.h"
+#include "SampleFrame.h"
+#include "ThreadPool.h"
+
+namespace lmms {
+SamplePreviewPlayHandle::SamplePreviewPlayHandle(const std::filesystem::path& path)
+	: PlayHandle(Type::SamplePlayHandle)
+{
+#ifdef LMMS_BUILD_WIN32
+	m_sndfile = sf_wchar_open(path.c_str(), SFM_READ, &m_sfinfo);
+#else
+	m_sndfile = sf_open(path.c_str(), SFM_READ, &m_sfinfo);
+#endif
+
+	if (m_sndfile == nullptr || m_sfinfo.channels == 0 || m_sfinfo.frames == 0)
+	{
+		throw std::runtime_error{"Failed to create sample preview stream"};
+	}
+
+	const auto initialFrameCount = std::min(static_cast<sf_count_t>(m_sfinfo.samplerate), m_sfinfo.frames);
+	m_buffer.resize((initialFrameCount + 1) * m_sfinfo.channels);
```

Comment:
Why are you comparing the sample rate with the frame count? And adding 1 to `initialFrameCount`?

Reply:
> Why are you comparing the sample rate with the frame count?

The code pulls a number of audio frames equal to the sample rate or the whole file itself, whichever is smaller. Pulling at most `sample rate` audio frames gives the non real-time writer thread enough leeway in case something goes wrong and it can't read into the buffer on time.

> And adding 1 to `initialFrameCount`?

The code adds 1 to the count because 1 of the slots in the buffer is for the empty slot. Note that the buffer is a circular buffer, and that empty slot allows us to distinguish between empty and full states of the buffer.

This could be a lot clearer however. I can make a simple `CircularBuffer` to dispel any confusion.

Reply:
Could you use `LocklessRingBuffer`?

Reply:
> Could you use `LocklessRingBuffer`?

I find `LocklessRingBuffer` to be a subpar implementation. It uses `QWaitCondition` which is not not the correct synchronization to use in real-time. The interface is also kind of odd as read functionality for `LocklessRingBuffer` should just be directly in there. A simple implementation that uses `std::atomic`and a simpler interface would be better IMO.

Reply:
Now that we have C++20's `std::atomic_flag` improvements, `std::counting_semaphore`, etc. I wonder if `LocklessRingBuffer` could be improved so that it doesn't need `QMutex` and `QWaitCondition`.